### PR TITLE
feat: endedness subcommand

### PR DESCRIPTION
**Pull request by a-frantz**
~Each input BAM returns two main values (in addition to evidence), which are titled `Mate state` (values of `Legal` or `Illegal`) and `Endedness` (values of `Paired-End`, `Single-End`, and `Inconclusive`).~
However the legal/illegal dichotomy does not agree with the SAM specification and we need new terms. I'm not sure what those should be, I'm hoping we can hash out the new terminology in this thread. Below is quoted the relevant pieces of the spec.

> Bits 0x40 and 0x80 reflect the read ordering within each template inherent in the sequencing
> technology used.^13 If 0x40 and 0x80 are both set, the read is part of a linear template, but it is
> neither the first nor the last read. If both 0x40 and 0x80 are unset, the index of the read in the
> template is unknown. This may happen for a non-linear template or when this information is lost
> during data processing.
>
> ^13: For example, in Illumina paired-end sequencing, first (0x40) corresponds to the R1 ‘forward’ read and last (0x80) to the R2 ‘reverse’ read. (Despite the terminology, this is unrelated to the segments’ orientations when they are mapped: either, neither, or both may have their reverse flag bits (0x10) set after mapping.)

Some anecdotal experience: I had easy access to a handful of SE BAM files on the cluster and processed them. Every read in those files did not have either read1/read2 bit set. A small sample size, but it implies the standard processing of SE data is to leave these bits unset. That contradicts the assumption I made (that SE data would have the read1 bit set) while formalizing this command's logic.
At the time of writing this comment, those files would return `Illegal \t Single-End`. I think the logic needs to be updated to make that return whatever we want to use instead of `Legal`.

As for the new term, it needs to convey two things:
1) any combination of setting or unsetting these bits across a file is technically allowed by the spec
2) we as "data processors" have expectations about what's an appropriate setting of these bits
    * i.e. certain combinations will make us say "we don't know what this data is conveying"
    * Those combinations are "legal"/"valid"/"allowed"/etc. but we don't know how to interpret them without extra context.

I propose we use the terms `Expected`/`Unexpected`. I don't love those, but they convey 1) and 2).

---------------------------
**EDIT WITH NEW LOGIC:**
The new logic should be straightforward to follow; it's all encapsulated in [this function](https://github.com/stjudecloud/ngsderive/pull/124/files#diff-c5ecde599507a56ad14b89257208332a50ce2f05c14749309426582bd0548d22R14).

But to put it into words and make explicit:
* if both first and last read bit are set for every read in the file && there is 1 "read per template" (RPT)
    * then return `Single-End`
* if (approx. 50% of reads have only the first bit set && every other read has only the last bit set) && there are 2 RPT
    * then return `Paired-End`
* else return `Unknown`
    * the code could be simplified to just this statement, but I prefer to be more explicit about other existing cases (that we may want to return a different value for in the future)

### CODE REVIEW

File: ngsderive/__main__.py [hunk 1]
```diff
@@ -207,10 +207,46 @@ class SaneFormatter(
...
+        type=float,
+        help="Distance from 0.5 split between read1s and read2s allowed to be called 'Paired-End'.",
+        default=0.05,
+    )
+    endedness.add_argument(
+        "--lenient",
+        action="store_true",
+        default=False,
+        help="Return a zero exit code on inconclusive results or "
+        + "when illegal mate states are discovered",
```
Comment (a-frantz) [inline ngsderive/__main__.py:236]: Just making a comment to remind me this help text will need to be updated

**Comment by claymcleod** [comment]
> At the time of writing this comment, those files would return Illegal \t Single-End. I think the logic needs to be updated to make that return whatever we want to use instead of Legal.

The excerpt above is presumably from the SAM specification. If so, it makes clear that neither flag should be interpreted as "Unknown" right? If so, I think a value of "Legal\tUnknown" is probably sufficient in these cases. We will just have to document this known case.

**Comment by a-frantz** [comment]
> > At the time of writing this comment, those files would return Illegal \t Single-End. I think the logic needs to be updated to make that return whatever we want to use instead of Legal.
> 
> The excerpt above is presumably from the SAM specification. If so, it makes clear that neither flag should be interpreted as "Unknown" right? If so, I think a value of "Legal\tUnknown" is probably sufficient in these cases. We will just have to document this known case.

First, that sentence is no longer true about what the tool returns. After 2 commits I pushed this morning, these cases would return `Expected \t Single-End`, not `Illegal \t Single-End`.

Re-quoting relevant bit:
> If both 0x40 and 0x80 are unset, the index of the read in the template is unknown

I skimmed through [this thread](https://github.com/samtools/hts-specs/issues/73) on the `hts-specs` repo, and my conclusion is "the spec is ambiguous and common usage is inconsistent with what logically follows from the spec". So IMO we need to look at some more real world examples from different aligners and see how these bits are being used in practice. My Googling isn't turning up much that's concrete about how to "properly" use these bits.

I've tested 3 Single End BAMs (An Oxford Nanopore Ultralong, one from 10X Genomics, and a PacBio BAM. They're test files @adthrasher found for the workflows repo) and none of them had 0x40 or 0x80 set on any reads. We know they are Single-End, so we can confidently conclude _at least_ some SE aligners don't set those bits (In my reading of the BAM specification, it is silent about how these bits should be set for Single-End data, leaving the implementation up to aligner authors).

IMO it would be incorrect for us to return `Inconclusive`/`Unknown` on these cases, even *if* that's what the specification says it means. Based on the real-world data I have access to, if 0x40 and 0x80 are unset for all reads, we're looking at Single-End data; and that this is an expected format for Single-End data.

In the thread I linked above, someone argued that all Single-End data should be setting BOTH bits, to indicate they are "first and last, i.e. the only read". That makes a lot of sense to me to be honest, but someone replied indicating that logic isn't used by anyone else.

After doing this little bit of research, I feel much less confident about what this tool should be returning for all these different cases. I'd like to dig into some more real world data to get a sense of what the community consensus is.

**Comment by claymcleod** [comment]
That sounds fine about thinking on it more. I still wouldn't feel comfortable saying that all data that has no flags set is single-ended. We know that tools don't follow the spec as they should, and in the past, rather than catering to the way the tools work, we have followed the spec and discussed the ambiguities with the spec authors. I would recommend we take the same approach:

- Call it `Conclusive/Unknown` for now (not that this is different the _inconclusive_, which you said above).
- Bring it up with the spec maintainers to clarify what tools should be expected to do.

**Comment by claymcleod** [comment]
I would recommend combining the reads per template into the final determination. For example, it's possible within your current design that, if every read was duplicated and had both flags set, the reads per template would be 2 but the code would still determine it as single-ended. This should instead serve up an error, because single-ended data should only have one read per template. See the discussion in Slack to see how I combined these in my mind.

**Comment by a-frantz** [comment]
The new logic should be straightforward to follow; it's all encapsulated in [this function](https://github.com/stjudecloud/ngsderive/pull/124/files#diff-c5ecde599507a56ad14b89257208332a50ce2f05c14749309426582bd0548d22R14).

But to put it into words and make explicit:
* if both first and last read bit are set for every read in the file && there is one "read per template" (RPT)
    * then return `Single-End`
* if (approx. 50% of reads have only the first bit set && every other read has only the last bit set) && there are two RPT
    * then return `Paired-End`
* else return `Unknown`
    * the code could be simplified to just this statement, but I prefer to be more explicit about other existing cases (that we may want to return a different value for in the future)

**Review by claymcleod — APPROVED** [review]
LGTM.

**Pull request merged**